File: backend/src/main/java/todo/GoalDataService.java
package todo;

import java.util.stream.Stream;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@Repository
public class GoalDataService {

  @Autowired private JdbcTemplate jdbcTemplate;

  public GoalData getByGoalDataId(long goalDataId) {
    String sql =
        "SELECT * FROM goal_data WHERE goal_data_id=?";
    RowMapper<GoalData> rowMapper = new GoalDataRowMapper();
    GoalData goalData = jdbcTemplate.queryForObject(sql, rowMapper, goalDataId);
    return goalData;
  }

  public long nextId() {
    String sql = "SELECT max(gd.goal_data_id) FROM goal_data gd";
    Long maxId = jdbcTemplate.queryForObject(sql, Long.class);
    if(maxId == null) {
      return 0;
    } else {
      return maxId + 1;
    }
  }

  public void add(GoalData goalData) {
    goalData.goalDataId = nextId();
    // Add goalData
    String sql =
        "INSERT INTO goal_data values (?, ?, ?, ?, ?, ?, ?, ?, ?)";
    jdbcTemplate.update(
        sql,
        goalData.goalDataId,
        goalData.creationTime,
        goalData.creatorUserId,
        goalData.goalId,
        goalData.name,
        goalData.startTime,
        goalData.duration,
        goalData.hidden,
        goalData.active
    );
  }

  public boolean existsByGoalDataId(long goalDataId) {
    String sql = "SELECT count(*) FROM goal_data gd WHERE gd.goal_data_id=?";
    int count = jdbcTemplate.queryForObject(sql, Integer.class, goalDataId);
    return count != 0;
  }

  // Restrict goalDatas by
  public Stream<GoalData> query(
      Long goalDataId,
      Long creationTime,
      Long minCreationTime,
      Long maxCreationTime,
      Long creatorUserId,
      Long goalId,
      String name,
      String partialName,
      String description,
      String partialDescription,
      Long duration,
      Long minDuration,
      Long maxDuration,
      Boolean hidden,
      Boolean active,
      boolean onlyRecent,
      Long courseId,
      long offset,
      long count) {

    boolean nojoingoal = courseId == null;

    String sql =
        "SELECT gd.* FROM goal_data gd"
            + (!onlyRecent ? "" : " INNER JOIN (SELECT max(goal_data_id) id FROM goal_data GROUP BY goal_id) maxids ON maxids.id = gd.goal_data_id")
            + (nojoingoal ? "" : " INNER JOIN goal ses ON ses.goal_id = gd.goal_id")
            + " WHERE 1=1 "
            + (goalDataId   == null ? "" : " AND gd.goal_data_id = " + goalDataId)
            + (creationTime    == null ? "" : " AND gd.creation_time = " + creationTime)
            + (minCreationTime == null ? "" : " AND gd.creation_time > " + minCreationTime)
            + (maxCreationTime == null ? "" : " AND gd.creation_time < " + maxCreationTime)
            + (creatorUserId   == null ? "" : " AND gd.creator_id = " + creatorUserId)
            + (goalId       == null ? "" : " AND gd.goal_id = " + goalId)
            + (name            == null ? "" : " AND gd.name = " + Utils.escape(name))
            + (partialName     == null ? "" : " AND gd.name LIKE " + Utils.escape("%"+partialName+"%"))
            + (description            == null ? "" : " AND gd.description = " + Utils.escape(description))
            + (partialDescription     == null ? "" : " AND gd.description LIKE " + Utils.escape("%"+partialDescription+"%"))
            + (duration        == null ? "" : " AND gd.duration = " + duration)
            + (minDuration     == null ? "" : " AND gd.duration > " + minDuration)
            + (maxDuration     == null ? "" : " AND gd.duration < " + maxDuration)
            + (hidden          == null ? "" : " AND gd.hidden = " + hidden)
            + (active          == null ? "" : " AND gd.active = " + active)
            + (courseId        == null ? "" : " AND ses.course_id = " + courseId)
            + (" ORDER BY gd.goal_data_id")
            + (" LIMIT " + offset + ", " + count)
            + ";";

    RowMapper<GoalData> rowMapper = new GoalDataRowMapper();
    return this.jdbcTemplate.query(sql, rowMapper).stream();
  }
}

File: backend2/src/goal_data_service.rs
use super::todo_app_db_types::*;
use super::utils::current_time_millis;
use tokio_postgres::GenericClient;
use std::convert::TryInto;
use todo_app_service_api::request;

impl From<tokio_postgres::row::Row> for GoalData {
  // select * from goal_data order only, otherwise it will fail
  fn from(row: tokio_postgres::Row) -> GoalData {
    GoalData {
      goal_data_id: row.get("goal_data_id"),
      creation_time: row.get("creation_time"),
      creator_user_id: row.get("creator_user_id"),
      goal_id: row.get("goal_id"),
      name: row.get("name"),
      duration_estimate: row.get("duration_estimate"),
      time_utility_function_id: row.get("time_utility_function_id"),
      parent_goal_id: row.get("parent_goal_id"),
      status: (row.get::<_, i64>("status") as u8).try_into().unwrap(),
    }
  }
}

// TODO we need to figure out a way to make scheduled and unscheduled goals work better
pub async fn add(
  con: &mut impl GenericClient,
  creator_user_id: i64,
  goal_id: i64,
  name: String,
  duration_estimate: i64,
  time_utility_function_id: i64,
  parent_goal_id: Option<i64>,
  status: request::GoalDataStatusKind,
) -> Result<GoalData, tokio_postgres::Error> {
  let creation_time = current_time_millis();

  let goal_data_id = con.query_one(
    "INSERT INTO
     goal_data(
         creation_time,
         creator_user_id,
         goal_id,
         name,
         duration_estimate,
         time_utility_function_id,
         parent_goal_id,
         status
     )
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
     RETURNING goal_data_id
    ",
    &[
      &creation_time,
      &creator_user_id,
      &goal_id,
      &name,
      &duration_estimate,
      &time_utility_function_id,
      &parent_goal_id,
      &(status.clone() as i64)
    ],
  ).await?.get(0);

  // return goal_data
  Ok(GoalData {
    goal_data_id,
    creation_time,
    creator_user_id,
    goal_id,
    name,
    duration_estimate,
    time_utility_function_id,
    parent_goal_id,
    status,
  })
}

pub async fn get_by_goal_data_id(
  con: &mut impl GenericClient,
  goal_data_id: i64,
) -> Result<Option<GoalData>, tokio_postgres::Error> {
  let result = con
    .query_opt(
      "SELECT * FROM goal_data WHERE goal_data_id=$1",
      &[&goal_data_id],
    ).await?
    .map(|x| x.into());
  Ok(result)
}

pub async fn query(
  con: &mut impl GenericClient,
  props: todo_app_service_api::request::GoalDataViewProps,
) -> Result<Vec<GoalData>, tokio_postgres::Error> {
  let sql = [
    "SELECT gd.* FROM goal_data gd",
    " JOIN goal g ON gd.goal_id = g.goal_id",
    if props.only_recent {
        " INNER JOIN (SELECT max(goal_data_id) id FROM goal_data GROUP BY goal_id) maxids
        ON maxids.id = gd.goal_data_id"
    } else {
        ""
    },
    " WHERE 1 = 1",
    " AND ($1  == NULL OR gd.goal_data_id = $1)",
    " AND ($2  == NULL OR gd.creation_time = $2)",
    " AND ($3  == NULL OR gd.creation_time >= $3)",
    " AND ($4  == NULL OR gd.creation_time <= $4)",
    " AND ($5  == NULL OR gd.creator_user_id = $5)",
    " AND ($6  == NULL OR gd.goal_id = $6)",
    " AND ($7  == NULL OR gd.name = $7)",
    " AND ($8  == NULL OR gd.partial_name LIKE CONCAT('%',$8,'%'))",
    " AND ($9  == NULL OR gd.duration_estimate = $9)",
    " AND ($10 == NULL OR gd.duration_estimate >= $10)",
    " AND ($11 == NULL OR gd.duration_estimate <= $11)",
    " AND ($12 == NULL OR gd.time_utility_function_id = $12)",
    " AND ($13 == NULL OR gd.parent_goal_id = $13)",
    " AND ($14 == NULL OR gd.status = $14)",
    " AND ($15 == NULL OR g.goal_intent_id = $15)",
    " ORDER BY gd.goal_data_id",
    " LIMIT $16, $17",
  ]
  .join("");

  let stmnt = con.prepare(&sql).await?;

  let results = con
    .query(
        &stmnt,

        & [
        &props.goal_data_id,
        &props.creation_time,
        &props.min_creation_time,
        &props.max_creation_time,
        &props.creator_user_id,
        &props.goal_id,
        &props.name,
        &props.partial_name,
        &props.duration_estimate,
        &props.min_duration_estimate,
        &props.max_duration_estimate,
        &props.time_utility_function_id,
        &props.parent_goal_id,
        &props.status.map(|x| x as i64),
        &props.goal_intent_id,
        &props.offset,
        &props.count,
    ]).await?
    .into_iter()
    .map(|row| row.into())
    .collect();

  Ok(results)
}
